### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by alpeb**
This is the implementation of the [CI-health-metrics RFC](https://github.com/linkerd/rfc/blob/master/design/0002-ci-health-metrics.md). Please check the [README.md](https://github.com/linkerd/linkerd2-ci-metrics/blob/master/README.md) for more information.

In order to generate a report using fake data, run the test in verbose mode:
```
go test ./cmd -v > report.html
```
and point your browser to that file.

To run the real deal, do
```
GITHUB_TOKEN=xxx go run ./cmd > report.html
```
where `GITHUB_TOKEN` can be any personal access token. Note that the "Failed Tests per Workflow" charts will be minimal for now, given we haven't started to expose errors in the integration tests as annotations (pending linkerd/linkerd2#4363).

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+		return totalSuccesses * 100 / totalRuns, pairlist.RankByValue(successes, false)
+	} else {
+		return 0, pairlist.PairList{}
+	}
+}
+
+// processData retrieves all the CI success and error message metrics and
+// displays them in an index.html file
+func processData(jobs []JobRun, annotations []ErrorAnn) error {
+	jobSuccessRatesJson, err := getJobSuccessRates(jobs)
```
Comment (kleimkuhler) [inline cmd/main.go]: linting:
```suggestions
	jobSuccessRatesJSON, err := getJobSuccessRates(jobs)
```

File: cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+	messages := []WorkflowWithMessages{}
+	for workflow := range setWorkflows {
+		m := WorkflowWithMessages{
+			Id:       strings.ReplaceAll(workflow, " ", "-"),
+			Name:     workflow,
+			Messages: getWorkflowMessages(workflow, annotations),
+		}
+		messages = append(messages, m)
+	}
+	workflowsJson, err := json.Marshal(messages)
```
Comment (kleimkuhler) [inline cmd/main.go]: linting:
```suggestion
	workflowsJSON, err := json.Marshal(messages)
```

File: cmd/main.go [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+			totalSuccesses++
+			successes[run.Workflow]++
+		}
+	}
+	for workflow, num := range successes {
+		successes[workflow] = num * 100 / totalRunsPerJob[workflow]
+	}
+	if totalRuns > 0 {
+		return totalSuccesses * 100 / totalRuns, pairlist.RankByValue(successes, false)
+	} else {
```
Comment (kleimkuhler) [inline cmd/main.go]: linting:
```suggestion
    if totalRuns > 0 {
		return totalSuccesses * 100 / totalRuns, pairlist.RankByValue(successes, false)
	}

    return 0, pairlist.PairList{}
```

File: cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+	ctx                  context.Context
+	client               *github.Client
+	optBigListPage       = &github.ListOptions{PerPage: 100}
+	workflows            = map[int64]string{}
+	now                  = time.Now()
+	monthAgo             = now.AddDate(0, -1, 0)
+	throttle             = time.Tick(rateLimit)
+)
+
+type JobRun struct {
```
Comment (kleimkuhler) [inline cmd/main.go:47]: nit: Each of these types could use a small comment explaining what they are used for.

File: cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+	MainCss              template.CSS
+	JobSuccessRatesArr   template.JS
+	WorkflowsArr         template.JS
+	Start                string
+	End                  string
+	GlobalSuccessRate    int
+	WorkflowSuccessRates pairlist.PairList
+}
+
+func rateLimitedCall() {
```
Comment (kleimkuhler) [inline cmd/main.go]: Needed?

**Review by kleimkuhler — COMMENTED** [review]
Overall this looks great! I don't have any major blocking comments on the code, but I do have a few questions about the generated report.

I generated a report with real data and it worked as described in the `README`. There are stats for `Create GH Release`, but both `Linkerd website publish check` and `Helm Chart Deploy` have a 0% success rate which seems incorrect. Do you know what is happening here? I will attach my `report.html`

Also, what are your thoughts on not grouping 100% success rate jobs into `Other`? I think it could be a little surprsing if looking for a specific job, but I also understand you are probably not using ci-metrics to look at stats for a job that never gives you issues.

Report (I cannot upload `.html` hence the `.txt`, you'll want to resave it as `.html` to open in browser):  [report.txt](https://github.com/linkerd/linkerd2-ci-metrics/files/4630145/report.txt)

**Comment by alpeb** [comment]
Thank you very much for the review @kleimkuhler :+1: 

> but both `Linkerd website publish check` and `Helm Chart Deploy` have a 0% success rate which seems incorrect

The problem was simply that the x axis started from 50%, not 0%, so those jobs actually had 50% success rate. I've modified the chart so it starts from 0%, which should be clearer.

> Also, what are your thoughts on not grouping 100% success rate jobs into `Other`? I think it could be a little surprsing if looking for a specific job, but I also understand you are probably not using ci-metrics to look at stats for a job that never gives you issues.

Agreed, better to be able to see all listed out. I've pushed a change for that. I've also addressed your other comments in the code :+1:

**Review by kleimkuhler — APPROVED** [review]
> The problem was simply that the x axis started from 50%, not 0%, so those jobs actually had 50% success rate.

Great yep I see that now. Also see all the jobs listed out. Thanks for making the changes, it looks awesome. I'm excited to see to see this in action!

**Pull request merged**